Add overloads for C# Interface to use flat vector with `unsafe` to speedup passing rays

mariuszhermansdorfer opened the pull request:
This PR has a bunch of minor optimizations to the raycasting API:

- Add an overload for the Embree raytracer to accept float[] as input
- Use unsafe methods to iterate over arrays for faster conversions

Test results:
![image](https://user-images.githubusercontent.com/49192999/205515046-7bf206b4-0eb4-46d3-b9f1-eabe71d1ecd4.png)

cadop commented on src/Csharp/CMakeLists.txt:
```diff
@@ -69,7 +69,7 @@ set_target_properties(
 		PROPERTIES
 		VS_DOTNET_REFERENCES "${normal_references}"
-		VS_DOTNET_TARGET_FRAMEWORK_VERSION "v4.5"
+		VS_DOTNET_TARGET_FRAMEWORK_VERSION "v4.8"
```
I updated the build instructions to mention this .net version issue https://github.com/cadop/dhart/pull/43

I would prefer to wait a little on incrementing this framework version until I can test the built dll's in different contexts (e.g. Unity).  

Could you revert the cmake changes for this PR?

mariuszhermansdorfer replied:
Changed in the latest commit

cadop commented on src/Csharp/packages/base/src/NativeUtils.cs:
```diff
@@ -180,7 +181,25 @@ public static float[] FlattenVectorArray(IEnumerable<Vector3D> vectors)
        */
-		public static Vector3D[] FloatArrayToVectorArray(float[] float_array, bool[] result_array)
+
+        public static float[] FlattenVectorArrayUnsafe(Vector3D[] vectors)
```
Could you confirm that the unittests are passing after you made this change? I haven't looked at this much yet, but it seems to be changing the API?

mariuszhermansdorfer replied:
Ah, my bad. I wanted to paste this under this method:
https://github.com/cadop/dhart/blob/44d2d9a52a18c4355c226116522bd7be99150f1a/src/Csharp/packages/base/src/NativeUtils.cs#L137-L160

I'm keeping it untouched and just introducing another overload working with Vector3D[]
![image](https://user-images.githubusercontent.com/49192999/205091771-43c33db2-beb3-4daa-9257-cf6468e1c172.png)

I'll put it in the right place

cadop commented on src/Csharp/packages/raytracer/src/RayTracerNative.cs:
```diff
@@ -16,7 +16,7 @@ namespace DHARTAPI.RayTracing
 
 	/*! \param NativeMethods for the RayTracing namespace */
-	internal static class NativeMethods
+	internal sealed class NativeMethods
```
Why is this changed to `sealed`?

mariuszhermansdorfer replied:
Performance reasons. Unless you want to derive from a class, it's generally a good idea to seal classes in c#
https://www.youtube.com/watch?v=d76WWAD99Yo

cadop replied:
I get the point he is making against regular classes, but we are using `static`. 

https://social.technet.microsoft.com/wiki/contents/articles/21028.difference-between-static-class-sealed-class-and-abstract-class-in-c.aspx

**Static Class:** ... This class **cannot be instantiated,** i.e we cannot have objects of this class. To access methods of this class, you can directly use classname.method. Also this class cannot be inherited.

**Sealed Class**:.... this class cannot be inherited. But **we can instantiate this class**, i.e we can have any number of objects of a sealed class.

So from the API standpoint, is there a reason you need to instantiate the class? I think this might be dangerous in the way we are interfacing with C, but i'm not sure. 


https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/static-classes-and-static-class-members

> Static classes are sealed and therefore cannot be inherited.

mariuszhermansdorfer replied:
Good point!
I’ll revert these back to static.

cadop commented:
Is there something in your gitignore or a missing commit? I can't build this branch

![image](https://user-images.githubusercontent.com/11399119/205106877-0344499f-4d4e-4ce4-9bc2-12fce0fc0b27.png)

mariuszhermansdorfer commented:
Try building it now @cadop.

I reverted all classes back to static and added the missing reference to CMakeLists. Previously I had it added manually here:
![image](https://user-images.githubusercontent.com/49192999/205131760-6b863ec3-47b6-4343-8de8-5b52a6409796.png)

cadop commented:
When you get a chance could you edit the original comment for this PR and paste a screenshot of the `hfcsharptest` group all passing.  I need to setup the automated testing on github runners, but until that happens, i'll just need to have people post that the unittests pass, and then ill review the code and test on other platforms.

mariuszhermansdorfer commented:
> When you get a chance could you edit the original comment for this PR and paste a screenshot of the `hfcsharptest` group all passing. I need to setup the automated testing on github runners, but until that happens, i'll just need to have people post that the unittests pass, and then ill review the code and test on other platforms.

Updated the initial comment and added a screenshot of all tests passing.

cadop commented:
Didn't mean to close this, it was because I deleted the other branch when merging it. You can rebase on main and reopen this PR to be merged into main (and ill change it).  

* You shouldn't need to make a new PR, just reopen this one
